File: server/datasets/python_source-code/Multiple-G2/structural_28_avl_tree.py
class Node_3:
    def __init__(self, v_3):
        self.v_3 = v_3
        self.r_3 = None
        self.l_3 = None
        self.height = 1

def h_3(node):
    if not node:
        return 0
    return node.height

def b_3(node):
    if not node:
        return 0
    return h_3(node.l_3) - h_3(node.r_3)

def rr_3(y):
    x = y.l_3
    sub = x.r_3
    x.r_3 = y
    y.l_3 = sub
    y.height = 1 + max(h_3(y.l_3), h_3(y.r_3))
    x.height = 1 + max(h_3(x.l_3), h_3(x.r_3))
    return x

def rl_3(x):
    y = x.r_3
    sub = y.l_3
    y.l_3 = x
    x.r_3 = sub
    x.height = 1 + max(h_3(x.l_3), h_3(x.r_3))
    y.height = 1 + max(h_3(y.l_3), h_3(y.r_3))
    return y

def ins_3(rt_3, v_3):
    if not rt_3:
        return Node_3(v_3)
    if v_3 < rt_3.v_3:
        rt_3.l_3 = ins_3(rt_3.l_3, v_3)
    elif v_3 > rt_3.v_3:
        rt_3.r_3 = ins_3(rt_3.r_3, v_3)
    else:
        return rt_3

    rt_3.height = 1 + max(h_3(rt_3.l_3), h_3(rt_3.r_3))
    bf_3 = b_3(rt_3)

    if bf_3 > 1 and v_3 < rt_3.l_3.v_3:
        return rr_3(rt_3)
    if bf_3 < -1 and v_3 > rt_3.r_3.v_3:
        return rl_3(rt_3)
    if bf_3 > 1 and v_3 > rt_3.l_3.v_3:
        rt_3.l_3 = rl_3(rt_3.l_3)
        return rr_3(rt_3)
    if bf_3 < -1 and v_3 < rt_3.r_3.v_3:
        rt_3.r_3 = rr_3(rt_3.r_3)
        return rl_3(rt_3)

    return rt_3

File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_g2/structural_28_avl_tree.java
class Node_3 {
    int k_3, h_3;
    Node_3 left, right;
    Node_3(int d) {
        k_3 = d;
        h_3 = 1;
    }
}
public class Tree_3 {
    public static void main(String[] args) {
        Tree_3 t = new Tree_3();
        Node_3 r = null;
        r = t.ins_3(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_3(Node_3 N) {
        if (N == null) return 0;
        return N.h_3;
    }
    int gb_3(Node_3 N) {
        if (N == null) return 0;
        return h_3(N.left) - h_3(N.right);
    }
    Node_3 upd_3(Node_3 n_3) {
        n_3.h_3 = 1 + max(h_3(n_3.left), h_3(n_3.right));
        return n_3;
    }
    Node_3 rr_3(Node_3 y) {
        Node_3 x = y.left;
        Node_3 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_3(y);
        upd_3(x);
        return x;
    }
    Node_3 rl_3(Node_3 x) {
        Node_3 y = x.right;
        Node_3 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_3(x);
        upd_3(y);
        return y;
    }
    Node_3 ins_3(Node_3 n_3, int k_3) {
        if (n_3 == null) return (new Node_3(k_3));
        if (k_3 < n_3.k_3) n_3.left = ins_3(n_3.left, k_3);
        else if (k_3 > n_3.k_3) n_3.right = ins_3(n_3.right, k_3);
        else return n_3;
        upd_3(n_3);
        int b_3 = gb_3(n_3);
        if (b_3 > 1 && k_3 < n_3.left.k_3) return rr_3(n_3);
        if (b_3 < -1 && k_3 > n_3.right.k_3) return rl_3(n_3);
        if (b_3 > 1 && k_3 > n_3.left.k_3) {
            n_3.left = rl_3(n_3.left);
            return rr_3(n_3);
        }
        if (b_3 < -1 && k_3 < n_3.right.k_3) {
            n_3.right = rr_3(n_3.right);
            return rl_3(n_3);
        }
        return n_3;
    }
}